### PR TITLE
FIX+DOC+STYLE: Improve code and docstring quality for v2.1.0 release

### PR DESCRIPTION
**Pull request by sebastientourbier**
This PR is dedicated to improve code and docstrings quality prior to the `v2.1.0` release.

As of commit https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/commit/e5626f6e8002eb7fc97ef0fd1377e896ec365658, It appears that:

- a number of docstrings are not complete or do not comply to the numpy docstring convention (https://numpydoc.readthedocs.io/en/latest/format.html) adopted in MIALSRTK. This results in an empty API documentation for the preprocess module as an example: https://mialsrtk.readthedocs.io/en/latest/api_preprocess.html. In addition, a number of interfaces do not provide example of usage. PR and review are in place to mitigate those risks.

- a number of unused imports are still present

- uncomment job to publish singularity image in circleci

### CODE REVIEW

**Comment by sebastientourbier** [comment]
Work still in progress

**Comment by sebastientourbier** [comment]
@t-sanchez I am sorry to tell you but PR #200 was not mature enough to be merged. MIALSRTK is developped following well-established conventions to ensure the perenity of code and documentation quality.

**Comment by t-sanchez** [comment]
Hey Sebastien, 

I noticed that you got the following error in circleCI 
> ImportError: dlopen: cannot load any more object with static TLS. Module "sklearn" could not be found. See http://nilearn.github.io/introduction.html#installation for installation information.

I remember having gotten this error after removing some paths that codacy suggested as unused but that seem to make the pipeline crash.  You might want to look into `srr.py` or other files where codacy suggested to remove imports. But maybe upgrading nipype could indeed solve the issue.

**Comment by sebastientourbier** [comment]
Thanks for the hint @t-sanchez  :) I indeed updated nipype to see if this would work with a new installation of the conda environment as the step `RUN micromamba install -v -y -n base -f /app/environment.yml` was using the cache. If sklearn is well installed, and this still does not work, I check the changes I made in `srr.py` between  https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/pull/202/commits/ac703ed35e18267d7bddbc42755322e7059b7aab and https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/pull/202/commits/671f520cacf0e1506f0c5c678c1955a9edca897a

**Comment by sebastientourbier** [comment]
Taking inspiration from this similar pytorch/sklearn [issue](https://github.com/scikit-learn/scikit-learn/issues/14485), this might be a bad interaction between versions of nilearn and sklearn.

**Comment by sebastientourbier** [comment]
> Taking inspiration from this similar pytorch/sklearn [issue](https://github.com/scikit-learn/scikit-learn/issues/14485), this might be a bad interaction between versions of nilearn and sklearn. I just updated nilearn to 0.10.0 which needs sklearn>=1.0.

It did not work, trying to downgrade sklearn this time...

**Comment by sebastientourbier** [comment]
Downgrading sklearn to 0.20.3 led to the following error:
```output
ImportError: A sklearn version of at least 0.21 is required to use nilearn. 0.20.3 was found. Please upgrade sklearn
```
Trying with 0.21

**Comment by t-sanchez** [comment]
I was checking the build of the documentation at [https://readthedocs.org/api/v2/build/19191965.txt](https://readthedocs.org/api/v2/build/19191965.txt) and I noticed the following warning:
```
WARNING: autodoc: failed to import module 'postprocess' from module 'pymialsrtk.interfaces'; the following exception was raised:
No module named 'SimpleITK'
WARNING: autodoc: failed to import module 'preprocess' from module 'pymialsrtk.interfaces'; the following exception was raised:
No module named 'nsol'
../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkImageReconstruction:1: WARNING: Explicit markup ends without a blank line; unexpected unindent.
../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkImageReconstruction:39: WARNING: Definition list ends without a blank line; unexpected unindent.
../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkSDIComputation:1: WARNING: Explicit markup ends without a blank line; unexpected unindent.
../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkTVSuperResolution:1: WARNING: Explicit markup ends without a blank line; unexpected unindent.
../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkTVSuperResolution.m_output_dict:1: WARNING: Duplicate explicit target name: "(link to paper)".
WARNING: autodoc: failed to import module 'anatomical' from module 'pymialsrtk.pipelines'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'anatomical.abstract' from module 'pymialsrtk.pipelines'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'anatomical.preprocessing' from module 'pymialsrtk.pipelines'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'anatomical.srr' from module 'pymialsrtk.pipelines'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'input_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'output_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
No module named 'SimpleITK'
WARNING: autodoc: failed to import module 'postproc_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'preproc_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'recon_labelmap_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'recon_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
No module named 'nsol'
WARNING: autodoc: failed to import module 'srr_assessment_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
No module named 'SimpleITK'
```
Do you think that this could be the cause of the documentation failing to generate?

**Comment by sebastientourbier** [comment]
Hello @t-sanchez , what happened to the `create_subject_report()`? I saw it is commented in srr.py but could not find its implementation anymore.

**Comment by t-sanchez** [comment]
Hey @sebastientourbier, we moved it to a dedicated nipype [node](https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/blob/e080709eb72bcdc03d439509a48fc5a965dbeb86/pymialsrtk/interfaces/postprocess.py#L1133).

**Comment by sebastientourbier** [comment]
So I guess I could delete the commented lines. Please, could you confirm?

**Comment by sebastientourbier** [comment]
> I was checking the build of the documentation at https://readthedocs.org/api/v2/build/19191965.txt and I noticed the following warning:
> 
> ```
> WARNING: autodoc: failed to import module 'postprocess' from module 'pymialsrtk.interfaces'; the following exception was raised:
> No module named 'SimpleITK'
> WARNING: autodoc: failed to import module 'preprocess' from module 'pymialsrtk.interfaces'; the following exception was raised:
> No module named 'nsol'
> ../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkImageReconstruction:1: WARNING: Explicit markup ends without a blank line; unexpected unindent.
> ../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkImageReconstruction:39: WARNING: Definition list ends without a blank line; unexpected unindent.
> ../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkSDIComputation:1: WARNING: Explicit markup ends without a blank line; unexpected unindent.
> ../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkTVSuperResolution:1: WARNING: Explicit markup ends without a blank line; unexpected unindent.
> ../pymialsrtk/interfaces/reconstruction.py:docstring of pymialsrtk.interfaces.reconstruction.MialsrtkTVSuperResolution.m_output_dict:1: WARNING: Duplicate explicit target name: "(link to paper)".
> WARNING: autodoc: failed to import module 'anatomical' from module 'pymialsrtk.pipelines'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'anatomical.abstract' from module 'pymialsrtk.pipelines'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'anatomical.preprocessing' from module 'pymialsrtk.pipelines'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'anatomical.srr' from module 'pymialsrtk.pipelines'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'input_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'output_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
> No module named 'SimpleITK'
> WARNING: autodoc: failed to import module 'postproc_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'preproc_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'recon_labelmap_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'recon_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
> No module named 'nsol'
> WARNING: autodoc: failed to import module 'srr_assessment_stage' from module 'pymialsrtk.workflows'; the following exception was raised:
> No module named 'SimpleITK'
> ```
> 
> @sebastientourbier, do you think that this could be the cause of the documentation failing to generate?

Sorry did not see this... Yes, this might be one of the reasons. 
I guess we can put nsol and SimpleITK in the following list of `documentation/conf.py`:
https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/blob/e5626f6e8002eb7fc97ef0fd1377e896ec365658/documentation/conf.py#L66-L76
I guess we do not need them for the documentation itself. If they are needed, they can be added to the `documentation/requirements.txt` file.

**Comment by t-sanchez** [comment]
> So I guess I could delete the commented lines. Please, could you confirm?
@sebastientourbier  Yes definitely :)

**Comment by t-sanchez** [comment]
> @t-sanchez If needed, I activated the generation of the docs for this branch on rtd. It can be viewed at the following URL:
https://mialsrtk.readthedocs.io/en/fix-code-and-docs-quality-for-v2.1.0-release/

@sebastientourbier Thanks :) I was checking through the automated builds. It seems that now the preprocess and postprocess modules are generated as intended.

**Comment by sebastientourbier** [comment]
Things I see that still need to be checked/updated:
- [ ] Check if substitution in data sinker for transforms is now working for tests on CircleCI.
- [ ] If yes, update the lists of test outputs
- [ ] API documentation of new stage workflows
- [ ] Update README on Dockerhub

**Comment by sebastientourbier** [comment]
@all-contributors please add @t-sanchez for bug, code, doc, example, ideas, infra

**Comment by sebastientourbier** [comment]
@t-sanchez do you see any contribution type I might have missed?

**Comment by sebastientourbier** [comment]
@all-contributors please add @t-sanchez for review. please add @pdedumast for design, ideas, review. please add @sebastientourbier for ideas, infra, mentoring.

**Comment by sebastientourbier** [comment]
@pdedumast @hamzake Do you see any contribution type(s) missing for you?

**Comment by t-sanchez** [comment]
@sebastientourbier This seems alright for me!

**Comment by t-sanchez** [comment]
Hi @sebastientourbier, I've checked with @hamzake and @pdedumast and it's fine also for them. 

I think that we're good to go!

**Comment by sebastientourbier** [comment]
Great! Thanks @t-sanchez! 
Yes I guess we are good to go!
I am going to take care of this tomorrow! 
Cheers

**Comment by sebastientourbier** [comment]
Hello @t-sanchez ! 
One last thing remaining. 
Please could you provide me with your ORCID? 
I added you in the `.zenodo.json` but still need to correct your ORCID in it.
https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/blob/8581ad47213b42fcae664f05ce66a9311021efe2/.zenodo.json#L30-L34
This file is used by Zenodo to create the DOI when we make the release on GitHub.

**Comment by t-sanchez** [comment]
Hey @sebastientourbier 

Here is my ORCID ID: 0000-0003-3668-5155

Regarding the affiliations, I think that it would be good to acknowledge, at least for Meri and I, the UNIL and CIBM affiliations as well. 
> Department of Radiology, Lausanne University Hospital (CHUV), University of Lausanne (UNIL),  and Center for Biomedical Imaging (CIBM)

I'll double check with Meri who needs which affiliation.

**Comment by t-sanchez** [comment]
Okay @sebastientourbier, the change is done. This is a bit wordy, so feel free to edit. We still need to keep all three affiliations + have CIBM in first for Meri.

**Review by sebastientourbier — COMMENTED** [review]
Seems ready to be merged in the `v2.1.0` branch for review!
